File: scripts/regenerate-local-landmarks.js
const fs = require("fs");
const path = require("path");
const https = require("https");

const ROOT = process.cwd();
const DATA_DIR = path.join(ROOT, "data");
const QUERY_PATH = path.join(DATA_DIR, "local-landmarks.overpassql");
const OVERPASS_JSON_PATH = path.join(DATA_DIR, "local-landmarks.overpass.json");
const GEOJSON_PATH = path.join(DATA_DIR, "local-landmarks.geojson");
const FOREST_BOUNDARY_PATH = path.join(DATA_DIR, "epping-forest-land.geojson");
const OVERPASS_URL = "https://overpass-api.de/api/interpreter";
const WALKING_SPEED_M_PER_MIN = 3500 / 60;
const MAX_WALK_MINUTES_FROM_BOUNDARY = 8;
const MAX_DISTANCE_FROM_BOUNDARY_METRES = WALKING_SPEED_M_PER_MIN * MAX_WALK_MINUTES_FROM_BOUNDARY;

function postOverpass(queryText) {
  return new Promise((resolve, reject) => {
    const requestBody = `data=${encodeURIComponent(queryText)}`;
    const req = https.request(
      OVERPASS_URL,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded; charset=utf-8",
          "Content-Length": Buffer.byteLength(requestBody),
        },
      },
      (res) => {
        let body = "";
        res.setEncoding("utf8");
        res.on("data", (chunk) => {
          body += chunk;
        });
        res.on("end", () => {
          if (res.statusCode < 200 || res.statusCode >= 300) {
            reject(new Error(`Overpass HTTP ${res.statusCode}: ${body.slice(0, 400)}`));
            return;
          }
          resolve(body);
        });
      }
    );

    req.on("error", reject);
    req.write(requestBody);
    req.end();
  });
}

function categoryLabel(key) {
  const labels = {
    pub: "Pub",
    bar: "Bar",
    cafe: "Café",
    tea: "Tea Hut",
    parking: "Car Park",
    toilets: "Toilets",
    bicycle_parking: "Cycle Parking",
    bench: "Bench",
    drinking_water: "Drinking Water",
    bus_station: "Bus Station",
    bus_stop: "Bus Stop",
    taxi: "Taxi",
    train_station: "Train Station",
    station: "Train Station",
    halt: "Rail Halt",
    tram_stop: "Tram Stop",
    gate: "Gate",
    stile: "Stile",
    kissing_gate: "Kissing Gate",
    cattle_grid: "Cattle Grid",
    lift_gate: "Lift Gate",
    swing_gate: "Swing Gate",
    cycle_barrier: "Cycle Barrier",
    entrance: "Entrance",
    attraction: "Attraction",
    viewpoint: "Viewpoint",
    museum: "Museum",
    picnic_site: "Picnic Site",
    information: "Information",
    camp_site: "Camp Site",
    caravan_site: "Caravan Site",
    historic: "Historic Site",
    monument: "Monument",
    archaeological_site: "Archaeological Site",
    memorial: "Memorial",
    ruins: "Ruins",
    castle: "Castle",
  };
  return labels[key] || String(key || "place").replace(/_/g, " ").replace(/\b\w/g, (m) => m.toUpperCase());
}

function polygonRings(geometry) {
  if (!geometry || !Array.isArray(geometry.coordinates)) return [];
  if (geometry.type === "Polygon") return geometry.coordinates;
  if (geometry.type === "MultiPolygon") {
    const rings = [];
    for (const polygon of geometry.coordinates) {
      if (Array.isArray(polygon)) rings.push(...polygon);
    }
    return rings;
  }
  return [];
}

function loadBoundarySegments() {
  const data = JSON.parse(fs.readFileSync(FOREST_BOUNDARY_PATH, "utf8"));
  const features = Array.isArray(data && data.features) ? data.features : [];
  const segments = [];
  for (const feature of features) {
    const rings = polygonRings(feature && feature.geometry);
    for (const ring of rings) {
      if (!Array.isArray(ring) || ring.length < 2) continue;
      for (let i = 1; i < ring.length; i += 1) {
        const a = ring[i - 1];
        const b = ring[i];
        if (!Array.isArray(a) || !Array.isArray(b) || a.length < 2 || b.length < 2) continue;
        segments.push([Number(a[0]), Number(a[1]), Number(b[0]), Number(b[1])]);
      }
      const first = ring[0];
      const last = ring[ring.length - 1];
      if (Array.isArray(first) && Array.isArray(last) && first.length >= 2 && last.length >= 2) {
        if (first[0] !== last[0] || first[1] !== last[1]) {
          segments.push([Number(last[0]), Number(last[1]), Number(first[0]), Number(first[1])]);
        }
      }
    }
  }
  return segments;
}

function toLocalXY(lon, lat, refLatRad) {
  return {
    x: lon * 111320 * Math.cos(refLatRad),
    y: lat * 110574,
  };
}

function pointToSegmentDistanceMetres(px, py, ax, ay, bx, by) {
  const abx = bx - ax;
  const aby = by - ay;
  const lengthSq = abx * abx + aby * aby;
  if (lengthSq <= 1e-12) return Math.hypot(px - ax, py - ay);
  let t = ((px - ax) * abx + (py - ay) * aby) / lengthSq;
  t = Math.max(0, Math.min(1, t));
  const cx = ax + t * abx;
  const cy = ay + t * aby;
  return Math.hypot(px - cx, py - cy);
}

function nearestBoundaryDistanceMetres(lon, lat, segments, refLatRad) {
  const point = toLocalXY(lon, lat, refLatRad);
  let best = Infinity;
  for (const [lon1, lat1, lon2, lat2] of segments) {
    const a = toLocalXY(lon1, lat1, refLatRad);
    const b = toLocalXY(lon2, lat2, refLatRad);
    const distance = pointToSegmentDistanceMetres(point.x, point.y, a.x, a.y, b.x, b.y);
    if (distance < best) best = distance;
  }
  return best;
}

function toFeature(element, boundarySegments, refLatRad) {
  const tags = element.tags || {};
  const longitude = typeof element.lon === "number" ? element.lon : element.center && element.center.lon;
  const latitude = typeof element.lat === "number" ? element.lat : element.center && element.center.lat;
  if (!Number.isFinite(longitude) || !Number.isFinite(latitude)) return null;

  const boundaryDistanceMetres = nearestBoundaryDistanceMetres(longitude, latitude, boundarySegments, refLatRad);
  if (boundaryDistanceMetres > MAX_DISTANCE_FROM_BOUNDARY_METRES) return null;

  const category = tags.amenity
    || tags.barrier
    || (tags.entrance ? "entrance" : null)
    || tags.railway
    || tags.highway
    || tags.tourism
    || tags.historic
    || (tags.heritage ? "historic" : null)
    || "place";
  const address = [
    [tags["addr:housenumber"], tags["addr:street"]].filter(Boolean).join(" "),
    [tags["addr:city"], tags["addr:postcode"]].filter(Boolean).join(" "),
  ]
    .filter(Boolean)
    .join(", ") || null;

  return {
    type: "Feature",
    id: `${element.type}/${element.id}`,
    geometry: {
      type: "Point",
      coordinates: [longitude, latitude],
    },
    properties: {
      id: `${element.type}/${element.id}`,
      osmType: element.type,
      osmId: element.id,
      name: tags.name || null,
      category,
      categoryLabel: categoryLabel(category),
      amenity: tags.amenity || null,
      tourism: tags.tourism || null,
      historic: tags.historic || null,
      railway: tags.railway || null,
      barrier: tags.barrier || null,
      entrance: tags.entrance || null,
      heritage: tags.heritage || null,
      website: tags.website || tags["contact:website"] || null,
      phone: tags.phone || tags["contact:phone"] || null,
      address,
      distanceToForestBoundaryMetres: Math.round(boundaryDistanceMetres * 10) / 10,
    },
  };
}

async function main() {
  const queryText = fs.readFileSync(QUERY_PATH, "utf8");
  const raw = await postOverpass(queryText);
  fs.writeFileSync(OVERPASS_JSON_PATH, raw);

  const parsed = JSON.parse(raw);
  const boundarySegments = loadBoundarySegments();
  if (!boundarySegments.length) {
    throw new Error(`No usable boundary segments in ${FOREST_BOUNDARY_PATH}`);
  }
  const latValues = boundarySegments.flatMap((segment) => [segment[1], segment[3]]).filter(Number.isFinite);
  const refLat = latValues.reduce((sum, value) => sum + value, 0) / Math.max(1, latValues.length);
  const refLatRad = (refLat * Math.PI) / 180;

  const features = (parsed.elements || []).map((element) => toFeature(element, boundarySegments, refLatRad)).filter(Boolean);

  const geojson = {
    type: "FeatureCollection",
    name: "Local landmarks around Epping Forest",
    generatedAt: new Date().toISOString(),
    source: {
      name: "OpenStreetMap via Overpass API",
      license: "Open Data Commons Open Database License (ODbL)",
      queryFile: "data/local-landmarks.overpassql",
      bbox: [-0.035, 51.595, 0.145, 51.745],
      distanceFilter: {
        maxMinutesFromForestBoundary: MAX_WALK_MINUTES_FROM_BOUNDARY,
        walkingSpeedMPerMin: WALKING_SPEED_M_PER_MIN,
        maxDistanceMetres: Math.round(MAX_DISTANCE_FROM_BOUNDARY_METRES * 10) / 10,
        boundaryFile: "data/epping-forest-land.geojson",
      },
    },
    features,
  };

  fs.writeFileSync(GEOJSON_PATH, JSON.stringify(geojson, null, 2) + "\n");

  const counts = {};
  for (const feature of features) {
    const key = feature.properties.category || "unknown";
    counts[key] = (counts[key] || 0) + 1;
  }

  console.log(`Wrote ${features.length} features to ${path.relative(ROOT, GEOJSON_PATH)}`);
  console.log("Category counts:");
  Object.keys(counts)
    .sort((a, b) => counts[b] - counts[a])
    .forEach((key) => {
      console.log(`  ${key}: ${counts[key]}`);
    });
}

main().catch((error) => {
  console.error(error.message || error);
  process.exit(1);
});

File: scripts/regenerate_local_landmarks.py
import json
import urllib.request
import urllib.parse
import urllib.error
from datetime import datetime, timezone
from pathlib import Path
import math

ROOT = Path.cwd()
DATA = ROOT / "data"
QUERY_PATH = DATA / "local-landmarks.overpassql"
OVERPASS_JSON_PATH = DATA / "local-landmarks.overpass.json"
GEOJSON_PATH = DATA / "local-landmarks.geojson"
FOREST_BOUNDARY_PATH = DATA / "epping-forest-land.geojson"
WALKING_SPEED_M_PER_MIN = 3500 / 60  # 3.5 km/h
MAX_WALK_MINUTES_FROM_BOUNDARY = 25
MAX_DISTANCE_FROM_BOUNDARY_METRES = WALKING_SPEED_M_PER_MIN * MAX_WALK_MINUTES_FROM_BOUNDARY
OVERPASS_URLS = [
    "https://overpass-api.de/api/interpreter",
    "https://overpass.kumi.systems/api/interpreter",
    "https://lz4.overpass-api.de/api/interpreter",
]

LABELS = {
    "pub": "Pub",
    "bar": "Bar",
    "cafe": "Café",
    "tea": "Tea Hut",
    "restaurant": "Restaurant",
    "parking": "Car Park",
    "toilets": "Toilets",
    "bicycle_parking": "Cycle Parking",
    "bench": "Bench",
    "drinking_water": "Drinking Water",
    "bus_station": "Bus Station",
    "bus_stop": "Bus Stop",
    "taxi": "Taxi",
    "train_station": "Train Station",
    "station": "Train Station",
    "halt": "Rail Halt",
    "tram_stop": "Tram Stop",
    "gate": "Gate",
    "stile": "Stile",
    "kissing_gate": "Kissing Gate",
    "cattle_grid": "Cattle Grid",
    "lift_gate": "Lift Gate",
    "swing_gate": "Swing Gate",
    "cycle_barrier": "Cycle Barrier",
    "entrance": "Entrance",
    "attraction": "Attraction",
    "viewpoint": "Viewpoint",
    "museum": "Museum",
    "picnic_site": "Picnic Site",
    "information": "Information",
    "camp_site": "Camp Site",
    "caravan_site": "Caravan Site",
    "historic": "Historic Site",
    "monument": "Monument",
    "archaeological_site": "Archaeological Site",
    "memorial": "Memorial",
    "ruins": "Ruins",
    "castle": "Castle",
}

def category_label(key: str) -> str:
    if key in LABELS:
        return LABELS[key]
    return (key or "place").replace("_", " ").title()


def polygon_rings(geometry):
    if not geometry:
        return []
    geom_type = geometry.get("type")
    coords = geometry.get("coordinates")
    if not isinstance(coords, list):
        return []
    if geom_type == "Polygon":
        return coords
    if geom_type == "MultiPolygon":
        rings = []
        for polygon in coords:
            if isinstance(polygon, list):
                rings.extend(polygon)
        return rings
    return []


def load_boundary_segments(path):
    if not path.exists():
        return []
    data = json.loads(path.read_text(encoding="utf-8"))
    features = data.get("features") if isinstance(data, dict) else []
    if not isinstance(features, list):
        return []
    segments = []
    for feature in features:
        geometry = (feature or {}).get("geometry") or {}
        for ring in polygon_rings(geometry):
            if not isinstance(ring, list) or len(ring) < 2:
                continue
            for i in range(1, len(ring)):
                a = ring[i - 1]
                b = ring[i]
                if isinstance(a, list) and isinstance(b, list) and len(a) >= 2 and len(b) >= 2:
                    segments.append((float(a[0]), float(a[1]), float(b[0]), float(b[1])))
            first = ring[0]
            last = ring[-1]
            if first != last and len(first) >= 2 and len(last) >= 2:
                segments.append((float(last[0]), float(last[1]), float(first[0]), float(first[1])))
    return segments


def to_local_xy(lon, lat, ref_lat_rad):
    x = lon * 111320.0 * math.cos(ref_lat_rad)
    y = lat * 110574.0
    return x, y


def point_to_segment_distance_m(px, py, ax, ay, bx, by):
    abx = bx - ax
    aby = by - ay
    length_sq = abx * abx + aby * aby
    if length_sq <= 1e-12:
        return math.hypot(px - ax, py - ay)
    t = ((px - ax) * abx + (py - ay) * aby) / length_sq
    t = max(0.0, min(1.0, t))
    cx = ax + t * abx
    cy = ay + t * aby
    return math.hypot(px - cx, py - cy)


def nearest_boundary_distance_m(lon, lat, segments, ref_lat_rad):
    px, py = to_local_xy(lon, lat, ref_lat_rad)
    best = float("inf")
    for ax, ay, bx, by in segments:
        d = point_to_segment_distance_m(px, py, ax, ay, bx, by)
        if d < best:
            best = d
    return best


def project_segments(segments, ref_lat_rad):
    projected = []
    for lon1, lat1, lon2, lat2 in segments:
        ax, ay = to_local_xy(lon1, lat1, ref_lat_rad)
        bx, by = to_local_xy(lon2, lat2, ref_lat_rad)
        projected.append((ax, ay, bx, by))
    return projected


def within_boundary_distance(lon, lat, projected_segments, ref_lat_rad, threshold_m):
    px, py = to_local_xy(lon, lat, ref_lat_rad)
    best = float("inf")
    for ax, ay, bx, by in projected_segments:
        d = point_to_segment_distance_m(px, py, ax, ay, bx, by)
        if d < best:
            best = d
            if best <= threshold_m:
                return True, best
    return best <= threshold_m, best

query_text = QUERY_PATH.read_text(encoding="utf-8")
raw = None
last_error = None

for endpoint in OVERPASS_URLS:
    try:
        request = urllib.request.Request(
            endpoint,
            data=query_text.encode("utf-8"),
            headers={"Content-Type": "text/plain; charset=utf-8"},
            method="POST",
        )
        with urllib.request.urlopen(request, timeout=120) as response:
            raw = response.read().decode("utf-8")
            print(f"Fetched from {endpoint} via POST")
            break
    except urllib.error.HTTPError as error:
        last_error = error
        try:
            encoded = urllib.parse.urlencode({"data": query_text})
            with urllib.request.urlopen(f"{endpoint}?{encoded}", timeout=120) as response:
                raw = response.read().decode("utf-8")
                print(f"Fetched from {endpoint} via GET")
                break
        except Exception as inner_error:
            last_error = inner_error
            continue
    except Exception as error:
        last_error = error
        continue

if raw is None:
    raise RuntimeError(f"Failed to fetch Overpass data: {last_error}")

OVERPASS_JSON_PATH.write_text(raw, encoding="utf-8")
parsed = json.loads(raw)
boundary_segments = load_boundary_segments(FOREST_BOUNDARY_PATH)
if not boundary_segments:
    raise RuntimeError(f"No usable boundary segments in {FOREST_BOUNDARY_PATH}")

lat_values = [seg[1] for seg in boundary_segments] + [seg[3] for seg in boundary_segments]
ref_lat_rad = math.radians(sum(lat_values) / max(1, len(lat_values)))
projected_boundary_segments = project_segments(boundary_segments, ref_lat_rad)

features = []
for element in parsed.get("elements", []):
    tags = element.get("tags", {})
    lon = element.get("lon")
    lat = element.get("lat")
    if lon is None or lat is None:
        center = element.get("center") or {}
        lon = center.get("lon")
        lat = center.get("lat")

    if not isinstance(lon, (int, float)) or not isinstance(lat, (int, float)):
        continue

    category = (
        tags.get("amenity")
        or tags.get("barrier")
        or ("entrance" if tags.get("entrance") else None)
        or tags.get("highway")
        or tags.get("railway")
        or tags.get("tourism")
        or tags.get("historic")
        or ("historic" if tags.get("heritage") else None)
        or "place"
    )
    address = ", ".join(
        part for part in [
            " ".join(p for p in [tags.get("addr:housenumber"), tags.get("addr:street")] if p),
            " ".join(p for p in [tags.get("addr:city"), tags.get("addr:postcode")] if p),
        ]
        if part
    ) or None

    is_within_limit, boundary_distance_m = within_boundary_distance(
        lon,
        lat,
        projected_boundary_segments,
        ref_lat_rad,
        MAX_DISTANCE_FROM_BOUNDARY_METRES,
    )
    if not is_within_limit:
        continue

    features.append({
        "type": "Feature",
        "id": f"{element.get('type')}/{element.get('id')}",
        "geometry": {"type": "Point", "coordinates": [lon, lat]},
        "properties": {
            "id": f"{element.get('type')}/{element.get('id')}",
            "osmType": element.get("type"),
            "osmId": element.get("id"),
            "name": tags.get("name"),
            "category": category,
            "categoryLabel": category_label(category),
            "amenity": tags.get("amenity"),
            "highway": tags.get("highway"),
            "tourism": tags.get("tourism"),
            "historic": tags.get("historic"),
            "railway": tags.get("railway"),
            "barrier": tags.get("barrier"),
            "entrance": tags.get("entrance"),
            "heritage": tags.get("heritage"),
            "website": tags.get("website") or tags.get("contact:website"),
            "phone": tags.get("phone") or tags.get("contact:phone"),
            "address": address,
            "distanceToForestBoundaryMetres": round(boundary_distance_m, 1),
        },
    })

geojson = {
    "type": "FeatureCollection",
    "name": "Local landmarks around Epping Forest",
    "generatedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
    "source": {
        "name": "OpenStreetMap via Overpass API",
        "license": "Open Data Commons Open Database License (ODbL)",
        "queryFile": "data/local-landmarks.overpassql",
        "bbox": [-0.035, 51.595, 0.145, 51.745],
        "distanceFilter": {
            "maxMinutesFromForestBoundary": MAX_WALK_MINUTES_FROM_BOUNDARY,
            "walkingSpeedMPerMin": WALKING_SPEED_M_PER_MIN,
            "maxDistanceMetres": round(MAX_DISTANCE_FROM_BOUNDARY_METRES, 1),
            "boundaryFile": "data/epping-forest-land.geojson",
        },
    },
    "features": features,
}

GEOJSON_PATH.write_text(json.dumps(geojson, indent=2) + "\n", encoding="utf-8")

counts = {}
for feature in features:
    key = feature["properties"].get("category") or "unknown"
    counts[key] = counts.get(key, 0) + 1

print(f"Wrote {len(features)} features")
for key in sorted(counts, key=lambda k: counts[k], reverse=True):
    print(f"{key}: {counts[key]}")
